libbeat/publisher/pipeline/output_test.go

```diff
@@ -0,0 +1,181 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package pipeline
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/outputs"
+	"github.com/elastic/beats/v7/libbeat/publisher"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+)
+
+func TestPublish(t *testing.T) {
+	tests := map[string]struct {
+		client outputs.Client
+	}{
+		"client": {
+			&mockClient{},
+		},
+		"network_client": {
+			&mockNetworkClient{},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			rand.Seed(time.Now().UnixNano())
+
+			wqu := makeWorkQueue()
+			makeClientWorker(nilObserver, wqu, test.client)
+
+			numEvents := atomic.MakeInt(0)
+			var wg sync.WaitGroup
+			for batchIdx := 0; batchIdx <= randIntBetween(25, 200); batchIdx++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					batch := randomBatch(50, 150, wqu)
+
+					numEvents.Add(len(batch.Events()))
+
+					wqu <- batch
+				}()
+			}
+			wg.Wait()
+
+			// Give some time for events to be published
+			time.Sleep(time.Duration(numEvents.Load()*3) * time.Microsecond)
+
+			// Make sure that all events have eventually been published
+			c := test.client.(interface{ Published() int })
+			assert.Equal(t, numEvents.Load(), c.Published())
+		})
+	}
+}
+
+func TestPublishWithClose(t *testing.T) {
+	rand.Seed(time.Now().UnixNano())
+
+	wqu := makeWorkQueue()
+	client := &mockNetworkClient{}
+	worker := makeClientWorker(nilObserver, wqu, client)
+
+	numEvents := atomic.MakeInt(0)
+	var wg sync.WaitGroup
+	for batchIdx := 0; batchIdx <= randIntBetween(25, 200); batchIdx++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			batch := randomBatch(50, 150, wqu)
+
+			numEvents.Add(len(batch.Events()))
+
+			wqu <- batch
+		}()
+	}
+
+	// Close worker before all batches have had time to be published
+	err := worker.Close()
+	require.NoError(t, err)
+
+	remaining := numEvents.Load() - client.Published()
+	assert.Greater(t, remaining, 0)
+
+	// Start new worker to drain work queue
+	makeClientWorker(nilObserver, wqu, client)
+	wg.Wait()
+
+	// Give some time for events to be published
+	time.Sleep(time.Duration(remaining*3) * time.Microsecond)
+
+	// Make sure that all events have eventually been published
+	assert.Equal(t, numEvents.Load(), client.Published())
+}
+
+type mockClient struct{ published int }
```

Comment:
we run our tests with `-race`. The way this type is used we should either use an atomic or protect operations via mutex.

Reply:
Fixed in 33391da48.